### PR TITLE
[ADD] attachment_azure

### LINKED ISSUE #222: [ADD] attachment_azure
Hello,

We have a customer who will run Odoo in Azure using Azure Blob Storage (https://azure.microsoft.com/en-us/services/storage/blobs/) for the Odoo filestore.

So we are planning to create a new module `attachment_azure`, similar to `attachment_s3`.

Do you mind hosting the module in this repo?

Thank you.

### PR DESCRIPTION
**Pull request by max3903**
Closes #222 

Still missing #223 

CC @hiren-serpentcs

### CODE REVIEW

**Comment by yvaucher** [comment]
Nice! I'll check that next week.

**Comment by p-tombez** [comment]
Looks promising. Would you consider adding the same mechanic for the storage name as in #224 ?

**Comment by max3903** [comment]
@p-tombez It is already taken care of here: https://github.com/camptocamp/odoo-cloud-platform/pull/225/files#diff-2f8200ae96993037ee309fce2e5b6eaae1bcfabc8613bc549dfe2b46c29c6285R96

**Comment by p-tombez** [comment]
@max3903 I don't think it's the same. Here, we don't have any control on the name instead of having an environment variable like in #224 where you can actually choose a name and add a placeholder in it for the DB name

**Comment by max3903** [comment]
@p-tombez Azure does not have the concept of a bucket. You create a storage account and a shared access signature and you get:
- a connection string
- a signature
- a sas url

With those credentials, we create a container for each database. So the storage account is multi tenant.

Additional argument for this, when Odoo uses the filesystem, sysadmin don't have the control on the name of the directory.

**Comment by max3903** [comment]
@p-tombez 

```
$ cat ~/.config/rclone/rclone.conf
[remote]
type = azureblob
sas_url = https://<storageaccountname>.blob.core.*****.net/?<shared_access_signature>
$ rclone lsd remote:
          -1 2021-05-26 21:58:42        -1 backup
          -1 2021-05-22 17:20:16        -1 production-master
          -1 2021-05-27 02:13:57        -1 qa-backup
          -1 2021-05-22 01:52:30        -1 qa-test
          -1 2021-05-27 21:35:13        -1 test-20210527
          -1 2021-05-27 21:09:46        -1 test-backup
```

File: attachment_azure/models/ir_attachment.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        * ``AZURE_STORAGE_ACCOUNT_NAME``
+        * ``AZURE_STORAGE_ACCOUNT_URL``
+        * ``AZURE_STORAGE_ACCOUNT_KEY``
+
+        """
+        connect_str = os.environ.get("AZURE_STORAGE_CONNECTION_STRING")
+        account_name = os.environ.get("AZURE_STORAGE_ACCOUNT_NAME")
+        account_url = os.environ.get("AZURE_STORAGE_ACCOUNT_URL")
+        account_key = os.environ.get("AZURE_STORAGE_ACCOUNT_KEY")
+        if not (connect_str or account_name or account_url or account_key):
```
Comment (yvaucher) [inline attachment_azure/models/ir_attachment.py]: Looking at the error message, it seems we want to test either:

`connect_str` is provided.

OR

All three other variable are provided.

Current test will not throw the exception if for instance only `AZURE_STORAGE_ACCOUNT_URL` is provided.

```suggestion
        if not (connect_str or (account_name and account_url and account_key)):
```

File: attachment_azure/models/ir_attachment.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+                _logger.exception(
+                    "Error during the connection to Azure container using the Shared "
+                    "Access Signature (SAS)"
+                )
+                raise exceptions.UserError(str(error))
+        return blob_service_client
+
+    @api.model
+    def _get_azure_container(self):
+        running_env = os.environ.get("RUNNING_ENV", "dev")
```
Comment (yvaucher) [inline attachment_azure/models/ir_attachment.py:95]: Might be worth documenting the `RUNNING_ENV` variable in the README. Where it comes from and what is the impact on the container name.

File: attachment_azure/models/ir_attachment.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+    @api.model
+    def _store_file_read(self, fname, bin_size=False):
+        if fname.startswith("azure://"):
+            container_client = self._get_azure_container()
+            key = fname.replace("azure://", "", 1).lower()
+            try:
+                blob_client = container_client.get_blob_client(key)
+                read = blob_client.download_blob().readall()
+            except exceptions.UserError:
```
Comment (yvaucher) [inline attachment_azure/models/ir_attachment.py]: This catch seems suspicious, `container_client` and `blob_client` are azure lib objects, I strongly doubt it raises Odoo like exceptions.

File: attachment_azure/models/ir_attachment.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+            with io.BytesIO() as file:
+                blob_client = container_client.get_blob_client(key.lower())
+                file.write(bin_data)
+                file.seek(0)
+                filename = "azure://%s" % (key)
+                try:
+                    blob_client.upload_blob(file, blob_type="BlockBlob")
+                except ResourceExistsError:
+                    pass
+                except exceptions.UserError as error:
```
Comment (yvaucher) [inline attachment_azure/models/ir_attachment.py]: Is this `except` at the right level? Also seems suspicious, as I doubt `upload_blob` will ever raise such an exception.

File: attachment_azure/models/ir_attachment.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        if fname.startswith("azure://"):
+            container_client = self._get_azure_container()
+            key = fname.replace("azure://", "", 1).lower()
+            # delete the file only if it is on the current configured container
+            # otherwise, we might delete files used on a different environment
+            try:
+                blob_client = container_client.get_blob_client(key)
+                blob_client.delete_blob()
+                _logger.info("File %s deleted on the object storage" % (fname))
+            except exceptions.UserError:
```
Comment (yvaucher) [inline attachment_azure/models/ir_attachment.py]: same here

**Review by yvaucher — CHANGES_REQUESTED** [review]
Thanks for your work on this module

Here are few remarks on Exception handling and for the parameters.

**Comment by max3903** [comment]
@yvaucher Thanks for the review. I believe all the comments have been taken care of.

**Review by yvaucher — APPROVED** [review]

**Pull request merged**